File: Exercise Files/4 Sorting/MergeSort/mergesort2.c

#include <stdio.h>

void print_items(int* p_items, int size) {
    for (int i = 0; i < size; i++) {
        printf("%d ", p_items[i]);
    }
    printf("\r\n");
    
}

void merge_sort(int* p_items, int size) {
    if (size > 1) {
        int middle;
        int i, j, k;
        int * leftpart;
        int * rightpart;
        leftpart = (int *) malloc(sizeof(int)*middle);
        rightpart = (int *) malloc(sizeof(int)*(size-middle));

      
        middle = size / 2;
        
        memcpy(leftpart, p_items, sizeof(int)*middle); // Filling left leftpart
        memcpy(rightpart, &p_items[middle], sizeof(int)*(size-middle));
        
        merge_sort(leftpart, middle);           
        merge_sort(rightpart, size-middle);
        i = 0;
        j = 0;
        k = 0;
        
        while ((i < middle) && (j < size-middle)) {
            if (leftpart[i] < rightpart[j]) {
                p_items[k] = leftpart[i];
                i += 1;
            } else {
                p_items[k] = rightpart[j];
                j += 1;
            }
            k += 1; 
        }
        
        if (i < middle) {
            for (i ; i < middle; i++) {
                p_items[k] = leftpart[i];
                k += 1;
            }
        }
        else if (j < size-middle) {
                for (j ; j < size-middle; j++) {
                p_items[k] = leftpart[j];
                k += 1;
            }
        }
            
    }
    
}

int main()
{
    printf("Hello World\n\r");
    
    int items[10] = {6, 20, 8, 19, 56, 23, 87, 41, 49, 53};
    print_items(items, 10);

    
    
    merge_sort(items, 10);
    print_items(items, 10);

    return 0;
}

File: Exercise Files/4 Sorting/MergeSort/mergesort_2.py



def merge_sort(in_numbers):     
    if len(in_numbers) > 1:
        
        left_part = in_numbers[:len(in_numbers)//2]
        right_part = in_numbers[len(in_numbers)//2:]

        merge_sort(left_part)
        merge_sort(right_part)

        i = 0 #index for left
        j = 0 #index for right
        k = 0 #index for in_numbers

        while(i < len(left_part) and j < len(right_part)):
            if (left_part[i] < right_part[j]):
                in_numbers[k] = left_part[i]
                i += 1
            else:
                in_numbers[k] = right_part[j]
                j += 1
            k += 1

        if (i < len(left_part)):
            in_numbers[k:] = left_part[i:]
        elif (j < len(right_part)):
            in_numbers[k:] = right_part[j:]








if __name__ == '__main__':
    ls = [3,5,7,85,3,2,45,8]
    print(ls)


    merge_sort(ls)
    print(ls)

